common/src/main/java/com/gregtechceu/gtceu/api/data/worldgen/generator/indicators/SurfaceIndicatorGenerator.java

```diff
@@ -0,0 +1,187 @@
+package com.gregtechceu.gtceu.api.data.worldgen.generator.indicators;
+
+import com.gregtechceu.gtceu.api.data.chemical.material.Material;
+import com.gregtechceu.gtceu.api.data.worldgen.GTOreDefinition;
+import com.gregtechceu.gtceu.api.data.worldgen.WorldGeneratorUtils;
+import com.gregtechceu.gtceu.api.data.worldgen.generator.IndicatorGenerator;
+import com.gregtechceu.gtceu.api.data.worldgen.ores.GeneratedVeinMetadata;
+import com.gregtechceu.gtceu.api.data.worldgen.ores.OreIndicatorPlacer;
+import com.gregtechceu.gtceu.common.block.SurfaceRockBlock;
+import com.gregtechceu.gtceu.common.data.GTBlocks;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.codecs.RecordCodecBuilder;
+import com.tterrag.registrate.util.nullness.NonNullSupplier;
+import lombok.AllArgsConstructor;
+import net.minecraft.MethodsReturnNonnullByDefault;
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.Direction;
+import net.minecraft.core.SectionPos;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.util.RandomSource;
+import net.minecraft.util.StringRepresentable;
+import net.minecraft.util.valueproviders.ConstantFloat;
+import net.minecraft.util.valueproviders.ConstantInt;
+import net.minecraft.util.valueproviders.FloatProvider;
+import net.minecraft.util.valueproviders.IntProvider;
+import net.minecraft.world.level.ChunkPos;
+import net.minecraft.world.level.WorldGenLevel;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.Blocks;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.chunk.BulkSectionAccess;
+import net.minecraft.world.level.levelgen.Heightmap;
+import org.apache.commons.lang3.function.TriFunction;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.UnaryOperator;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+
+@MethodsReturnNonnullByDefault
+@ParametersAreNonnullByDefault
+public class SurfaceIndicatorGenerator extends IndicatorGenerator {
+    public static final Codec<SurfaceIndicatorGenerator> CODEC = RecordCodecBuilder.create(instance -> instance.group(
+            BuiltInRegistries.BLOCK.byNameCodec().fieldOf("block").forGetter(ext -> ext.block.get()),
```

Comment:
why not make this a blockstate instead of...

Reply:
or, another option: make it an `Either<Material, BlockState>` where if it's a material the dust is automatically assumed and the property is set.

Reply:
in short, you need a `Codec.either(BlockState.CODEC, GTRegistries.MATERIALS.codec())` and then resolve that at some point to a blockSTATE, which you can then directly place.

Reply:
That would actually also work rather well with the placement types.